### PR TITLE
Windows: Publish envoy-windows-debug container

### LINKED ISSUE #14231: Windows: Add Windows per-master commit debug container
Add an analogous Windows container image to https://hub.docker.com/r/envoyproxy/envoy-debug-dev now that we have a PDB file for envoy-static.exe

Action item from Windows dev community meeting

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Commit Message:
Windows: Publish envoy-windows-debug container

Same as release, but adds envoy.pdb
Additional Description: N/A
Risk Level: Low
Testing: Tested image build locally
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Fixes #14231

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
cc @envoyproxy/windows-dev

File: ci/Dockerfile-envoy-windows [hunk 1]
```diff
@@ -3,9 +3,12 @@ ARG BUILD_TAG=ltsc2019
 
 FROM $BUILD_OS:$BUILD_TAG
 
+ARG BUILD_TYPE=release
+
 RUN mkdir "C:\\Program\ Files\\envoy"
 RUN setx path "%path%;c:\Program Files\envoy"
-ADD ["windows/amd64/envoy.exe", "C:/Program Files/envoy/"]
+ADD ["windows/amd64/envoy.*", "C:/Program Files/envoy/"]
```
Comment (wrowe) [inline ci/Dockerfile-envoy-windows:10]: Only one concern, is this limited to envoy.exe and envoy.pdb? The linker often emits other crazy output such as .map and .lib files, even for .exe files.

File: ci/docker_ci.sh [hunk 2]
```diff
@@ -118,9 +119,8 @@ fi
 # This prefix is altered for the private security images on setec builds.
 DOCKER_IMAGE_PREFIX="${DOCKER_IMAGE_PREFIX:-envoyproxy/envoy}"
 
-
 if is_windows; then
-  BUILD_TYPES=("-windows")
+  BUILD_TYPES=("-windows" "-windows-debug")
```
Comment (wrowe) [inline ci/docker_ci.sh:123]: Question, does this cause the linker to run twice, once for windows and once for windows-debug? The issue is that the resulting .pdb file wouldn't be compatible with the resulting .exe file generally distributed to the public. There must be only one linked .exe, and the .pdb captured only for those who care to dive into crash dump files.

File: ci/Dockerfile-envoy-windows (hunk 1 above)
Comment (sunjayBhatia) [reply ci/Dockerfile-envoy-windows:10]: the `windows/amd64` directory will only contain what we put in the tarball we created here: https://github.com/envoyproxy/envoy/pull/14265/files#diff-c2b36673c33d9d04ccfde19e54e4a5273d69b47523a258f91de9c71e5ef87e61R76

(tarball downloaded here and untarred) https://github.com/envoyproxy/envoy/blob/d1ded6b381ca92cbacb2e0683adf997239b12272/.azure-pipelines/pipelines.yml#L418-L430

File: ci/docker_ci.sh (hunk 2 above)
Comment (sunjayBhatia) [reply ci/docker_ci.sh:123]: The docker image build grabs the build output tarball from the windows pipeline run it is in (the previous `windows.release` step) so it should have the `envoy.exe` and `envoy.pdb` file from that one `opt` build run

**Comment by sunjayBhatia** [comment]
hmm the failure `cp: cannot stat 'bazel-bin/source/exe/envoy-static.pdb': No such file or directory` might have to do something with RBE im thinking

**Comment by sunjayBhatia** [comment]
By default, Bazel is already downloading all target outputs: https://docs.bazel.build/versions/master/command-line-reference.html#flag--remote_download_outputs We need Bazel to consider the `.pdb` files as outputs of the `envoy_binary` rule

**Comment by sunjayBhatia** [comment]
Opened https://github.com/bazelbuild/rules_cc/issues/94 if anyone wants to chime in

File: ci/Dockerfile-envoy-windows [hunk 3]
```diff
@@ -3,9 +3,12 @@ ARG BUILD_TAG=ltsc2019
...
+RUN if "%BUILD_TYPE%" == "release" del "C:\Program Files\envoy\envoy.pdb"
```
Comment (lizan) [inline ci/Dockerfile-envoy-windows:11]: Won't this still leave envoy.pdb in docker layers if the image isn't squashed?

Comment (sunjayBhatia) [reply ci/Dockerfile-envoy-windows:11]: Yeah, I think you are right. I don’t know of a way to conditionally do an ADD, might have to change it to how we’ve done it on Linux with two different directories to ADD from between debug and regular image builds

Comment (lizan) [reply ci/Dockerfile-envoy-windows:11]: In Linux build we setup two directory and use directory name to change what file to add.

**Comment by sunjayBhatia** [comment]
Blocked for now on getting pdb output as part of a cc_binary rule, moving back to draft

File: ci/Dockerfile-envoy-windows (hunk 1 above)
Comment (wrowe) [reply ci/Dockerfile-envoy-windows:10]: The build artifact seems to include the tarball and the tree the tarball is created from, is this an issue?

**Comment by wrowe** [comment]
Unstale, in anticipation of eventual bazel handling of pdb output.

File: ci/Dockerfile-envoy-windows (hunk 1 above)
Comment (wrowe) [reply ci/Dockerfile-envoy-windows:10]: Moving this q... comment... etc over to discussions of windows_build_steps.sh

File: ci/windows_ci_steps.sh [hunk 4]
```diff
@@ -70,9 +70,10 @@ bazel "${BAZEL_STARTUP_OPTIONS[@]}" build "${BAZEL_BUILD_OPTIONS[@]}" //source/e
 
 # Copy binary to delivery directory
 cp -f bazel-bin/source/exe/envoy-static.exe "${ENVOY_DELIVERY_DIR}/envoy.exe"
+cp -f bazel-bin/source/exe/envoy-static.pdb "${ENVOY_DELIVERY_DIR}/envoy.pdb"
```
Comment (wrowe) [inline ci/windows_ci_steps.sh:74]: For anyone observing from the bleacher seats... various attempts to resolve this are noted in https://github.com/envoyproxy/envoy/pull/16043

**Comment by wrowe** [comment]
/wait

This is blocked on bazel release 4.2 for commit https://github.com/bazelbuild/bazel/commit/6dc941e58dfc1d4a9714a76b921fbe11fce658ed and needs a corresponding rules_cc bump.

**Comment by wrowe** [comment]
Note to selves @alyssawilk - waiting tag does not avert stalebot.

**Comment by wrowe** [comment]
@sunjayBhatia seems with 16043 merged, this can be woken up from stalled in the coming weeks.

**Comment by sunjayBhatia** [comment]
I'm unlikely to work on this any time soon, but whoever is interested can take this using this branch or start again!

**Pull request closed without merging**